File: experimental/typescript/src/core/bundle.ts
/**
 * Bundle context and configuration
 *
 * Provides access to bundle metadata and variable resolution during
 * resource loading and mutation.
 */

import { isVariable, type VariableOr, type VariableOrList } from "./variable.js";

/**
 * Bundle contains information about a bundle accessible in functions
 * loading and mutating resources.
 */
export class Bundle {
  /**
   * Selected target where the bundle is being loaded.
   * E.g.: 'development', 'staging', or 'production'.
   */
  readonly target: string;

  /**
   * Values of bundle variables resolved for selected target.
   * Bundle variables are defined in databricks.yml.
   *
   * For accessing variables as structured data, use resolveVariable().
   *
   * @example
   * ```yaml
   * variables:
   *   default_dbr_version:
   *     description: Default version of Databricks Runtime
   *     default: "14.3.x-scala2.12"
   * ```
   */
  readonly variables: Readonly<Record<string, unknown>>;

  /**
   * Bundle mode (e.g., "development", "production")
   * Set via the `mode` field in databricks.yml target configuration.
   */
  readonly mode?: string;

  /**
   * Bundle name from databricks.yml
   */
  readonly name: string;

  constructor(options: {
    target: string;
    variables?: Record<string, unknown>;
    mode?: string;
    name: string;
  }) {
    this.target = options.target;
    this.variables = Object.freeze({ ...options.variables });
    this.mode = options.mode;
    this.name = options.name;
  }

  /**
   * Returns true if the bundle is in development mode.
   *
   * @returns true if mode is "development", false otherwise
   *
   * @example
   * ```typescript
   * if (bundle.isDevelopment) {
   *   // Add dev-specific configuration
   * }
   * ```
   */
  get isDevelopment(): boolean {
    return this.mode === "development";
  }

  /**
   * Returns true if the bundle is in production mode.
   *
   * @returns true if mode is "production", false otherwise
   *
   * @example
   * ```typescript
   * if (bundle.isProduction) {
   *   // Add production-specific configuration
   * }
   * ```
   */
  get isProduction(): boolean {
    return this.mode === "production";
  }

  /**
   * Resolve a variable to its value.
   *
   * If the value is a variable, it will be resolved and returned.
   * Otherwise, the value will be returned as is.
   *
   * @throws Error if the variable is not found or cannot be resolved
   */
  resolveVariable<T>(variable: VariableOr<T>): T {
    if (!isVariable(variable)) {
      return variable;
    }

    const VAR_PREFIX = "var";

    if (!variable.path.startsWith(`${VAR_PREFIX}.`)) {
      throw new Error(`You can only get values of variables starting with '${VAR_PREFIX}.*'`);
    }

    const variableName = variable.path.slice(VAR_PREFIX.length + 1);

    if (!(variableName in this.variables)) {
      throw new Error(
        `Can't find '${variableName}' variable. Did you define it in databricks.yml?`
      );
    }

    const value = this.variables[variableName];

    // Check if the value is a nested variable reference
    const nestedPath = unwrapVariablePath(value);
    if (nestedPath) {
      // The value itself is a variable reference like "${var.other}"
      // Check if the variable type allows nested variables
      throw new Error(
        `Failed to resolve '${variableName}' because it refers to another variable '${nestedPath}'. ` +
          `Consider using Variable<VariableOr<T>> for nested variable support.`
      );
    }

    return value as T;
  }

  /**
   * Resolve a list variable to its value.
   *
   * If the value is a variable, or the list items are variables,
   * they will be resolved and returned.
   * Otherwise, the value will be returned as is.
   */
  resolveVariableList<T>(variable: VariableOrList<T>): T[] {
    const resolved = this.resolveVariable(variable);

    if (!Array.isArray(resolved)) {
      throw new Error("Expected a list value");
    }

    return resolved.map((item) => this.resolveVariable(item));
  }
}

/**
 * Checks if a value is a string that looks like a variable reference.
 * Returns the variable path if found, undefined otherwise.
 *
 * @param value - Value to check
 * @returns Variable path (e.g., "var.foo") or undefined
 */
function unwrapVariablePath(value: unknown): string | undefined {
  if (typeof value !== "string") {
    return undefined;
  }

  // Check if it matches ${var.something}
  const match = value.match(/^\$\{(.+)\}$/);
  return match ? match[1] : undefined;
}

File: python/databricks/bundles/core/_bundle.py
from dataclasses import dataclass, field
from typing import Any, TypeVar, Union, get_origin

from databricks.bundles.core._variable import Variable, VariableOr, VariableOrList

__all__ = [
    "Bundle",
]

_T = TypeVar("_T")

_VAR_PREFIX = "var"


@dataclass(frozen=True, kw_only=True)
class Bundle:
    """
    Bundle contains information about a bundle accessible in functions
    loading and mutating resources.
    """

    target: str
    """
    Selected target where the bundle is being loaded. E.g.: 'development', 'staging', or 'production'.
    """

    variables: dict[str, Any] = field(default_factory=dict)
    """
    Values of bundle variables resolved for selected target. Bundle variables are defined in databricks.yml.
    For accessing variables as structured data, use :meth:`resolve_variable`.

    Example:

    .. code-block:: yaml

        variables:
          default_dbr_version:
            description: Default version of Databricks Runtime
            default: "14.3.x-scala2.12"
    """

    _resource_state: dict[str, dict[str, dict[str, Any]]] = field(default_factory=dict, repr=False)
    """
    Internal field containing deployed resource state information.
    Use :meth:`get_resource_id` to access deployed resource IDs.
    """

    def get_resource_id(self, resource_type: str, resource_name: str) -> str | None:
        """
        Get the ID of a deployed resource.

        This method is available in post-deploy callbacks to retrieve deployed resource IDs.

        :param resource_type: The type of resource (e.g., 'jobs', 'pipelines', 'schemas')
        :param resource_name: The name of the resource as defined in the bundle
        :return: The deployed resource ID, or None if not found or not yet deployed

        Example:

        .. code-block:: python

            def on_deploy_complete(bundle: Bundle) -> None:
                job_id = bundle.get_resource_id("jobs", "my_job")
                if job_id:
                    print(f"Job deployed with ID: {job_id}")
        """
        resources = self._resource_state.get("resources", {})
        resource_group = resources.get(resource_type, {})
        resource_info = resource_group.get(resource_name, {})
        return resource_info.get("id")

    def resolve_variable(self, variable: VariableOr[_T]) -> _T:
        """
        Resolve a variable to its value.

        If the value is a variable, it will be resolved and returned.
        Otherwise, the value will be returned as is.
        """
        if not isinstance(variable, Variable):
            return variable

        if not variable.path.startswith(_VAR_PREFIX + "."):
            raise ValueError(
                "You can only get values of variables starting with 'var.*'"
            )
        else:
            variable_name = variable.path[len(_VAR_PREFIX + ".") :]

        if variable_name not in self.variables:
            raise ValueError(
                f"Can't find '{variable_name}' variable. Did you define it in databricks.yml?"
            )

        value = self.variables.get(variable_name)

        # avoid circular import
        from databricks.bundles.core._transform import (
            _display_type,
            _find_union_arg,
            _transform,
            _unwrap_variable_path,
        )

        if nested := _unwrap_variable_path(value):
            can_be_variable = get_origin(variable.type) == Union and _find_union_arg(
                nested, variable.type
            )
            can_be_variable = can_be_variable or get_origin(variable.type) == Variable

            if not can_be_variable:
                display_type = _display_type(variable.type)

                raise ValueError(
                    f"Failed to resolve '{variable_name}' because refers to another "
                    f"variable '{nested}'. Change variable type to "
                    f"Variable[VariableOr[{display_type}]]"
                )

        try:
            return _transform(variable.type, value)
        except Exception as e:
            raise ValueError(f"Failed to read '{variable_name}' variable value") from e

    def resolve_variable_list(self, variable: VariableOrList[_T]) -> list[_T]:
        """
        Resolve a list variable to its value.

        If the value is a variable, or the list item is a variable, it will be resolved and returned.
        Otherwise, the value will be returned as is.
        """

        return [self.resolve_variable(item) for item in self.resolve_variable(variable)]
